### PR TITLE
Follow up — changes to load model state in checkpoint connector in case of multiple workers #8044

### PR DESCRIPTION
**Pull request by mleshen**
## What does this PR do?

Fixes #<issue_number>

## Before submitting
- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [ ] Did you list all the **breaking changes** introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/fully_sharded.py [hunk 1]
```diff
@@ -178,6 +182,19 @@ def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
...
+        checkpoint = {}
+        log.info(f"FullyShardedDataParallel has {self.num_processes} processes. Serializing to avoid CPU OOMs.")
+        for current_worker in range(self.num_processes):
+            if self.local_rank == current_worker:
+                checkpoint = super().load_checkpoint_file(checkpoint_path)
+                self.lightning_module.on_load_checkpoint(checkpoint)
+                self.load_model_state_dict(checkpoint)
+                log.info(f"Rank {self.global_rank}: done loading model states from {checkpoint_path}.")
+                del checkpoint["state_dict"]
+            self.barrier()
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/fully_sharded.py:195]: this is pretty smart :) 

one concern, now the responsibility of calling the hooks is shifted to the plugin. do we want to allow that?
@PyTorchLightning/core-contributors

File: pytorch_lightning/plugins/training_type/fully_sharded.py [hunk 2]
```diff
@@ -178,6 +182,19 @@ def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
         # state dict.
         return super().lightning_module_state_dict()
 
+    def serialized_restore_model_state(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
+        checkpoint = {}
+        log.info(f"FullyShardedDataParallel has {self.num_processes} processes. Serializing to avoid CPU OOMs.")
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/fully_sharded.py:187]: we probably don't want to print that several times right?
so we could use the function `rank_zero_info` from the utilities.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -145,12 +145,20 @@ def restore_model(self) -> None:
     def restore_model_weights(self, checkpoint_path: Optional[Union[str, Path]]) -> None:
         """ Restore only the model weights. """
         checkpoint = self._loaded_checkpoint
+        if hasattr(self.trainer.training_type_plugin, "serialized_restore_model_state"):
+            checkpoint = self.trainer.training_type_plugin.serialized_restore_model_state(checkpoint_path)
+        else:
+            checkpoint = self.restore_model_state(checkpoint_path)
+
+    def restore_model_state(self, checkpoint_path: Optional[Union[str, Path]]) -> dict:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:153]: maybe it's better to just incorporate it into the above if else. 
a separate method with a similar name is maybe a bit confusing?

File: pytorch_lightning/plugins/training_type/fully_sharded.py (hunk 2 above)
Comment (mleshen) [reply pytorch_lightning/plugins/training_type/fully_sharded.py]: thanks for the suggestion; what a clever method!

File: pytorch_lightning/plugins/training_type/fully_sharded.py [hunk 4]
```diff
@@ -178,6 +182,25 @@ def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
         # state dict.
         return super().lightning_module_state_dict()
 
+    def serialized_restore_model_state(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
+        checkpoint = {}
+        rank_zero_info(
+            f"FullyShardedDataParallel has {self.num_processes} processes. Serializing model
+            state restore to avoid CPU OOMs."
+        )
+        for current_worker in range(self.num_processes):
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/fully_sharded.py:191]: Would this break if the number of gpus on checkpointing / saving isn't the same. Should we save the world_size + rank in each checkpoint and re-use that information on reload ?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 5]
```diff
@@ -146,10 +146,12 @@ def restore_model_weights(self, checkpoint_path: Optional[Union[str, Path]]) ->
         """ Restore only the model weights. """
         checkpoint = self._loaded_checkpoint
         if checkpoint_path is not None:
-            checkpoint = self.trainer.training_type_plugin.load_checkpoint_file(checkpoint_path)
-
-        self.trainer.lightning_module.on_load_checkpoint(checkpoint)
-        self.trainer.training_type_plugin.load_model_state_dict(checkpoint)
+            if hasattr(self.trainer.training_type_plugin, "serialized_restore_model_state"):
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:149]: Just personal preference, but I would prefer to refactor the code and pass the `checkpoint_path` to the training_type plugin and it handles the loading logic.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 6]
```diff
@@ -146,10 +146,12 @@ def restore_model_weights(self, checkpoint_path: Optional[Union[str, Path]]) ->
...
+                checkpoint = self.trainer.training_type_plugin.serialized_restore_model_state(checkpoint_path)
+            else:
+                checkpoint = self.trainer.training_type_plugin.load_checkpoint_file(checkpoint_path)
+                self.trainer.lightning_module.on_load_checkpoint(checkpoint)
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:153]: For consistency, should we move `self.trainer.lightning_module.on_load_checkpoint(checkpoint)` to training type plugin ?

File: pytorch_lightning/plugins/training_type/fully_sharded.py [hunk 7]
```diff
@@ -178,6 +182,25 @@ def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
...
+            if self.local_rank == current_worker:
+                checkpoint = super().load_checkpoint_file(checkpoint_path)
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/fully_sharded.py:193]: Does this assume the same checkpoint path for all workers ?

File: pytorch_lightning/plugins/training_type/fully_sharded.py [hunk 8]
```diff
@@ -178,6 +182,25 @@ def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
...
+                self.lightning_module.on_load_checkpoint(checkpoint)
+                self.load_model_state_dict(checkpoint)
+                del checkpoint["state_dict"]
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/fully_sharded.py:196]:
```suggestion
                self.load_model_state_dict(checkpoint.pop("state_dict"))
```

File: pytorch_lightning/plugins/training_type/fully_sharded.py (hunk 4 above)
Comment (mleshen) [reply pytorch_lightning/plugins/training_type/fully_sharded.py:191]: thanks for raising this! thinking aloud — in the same training session we wouldn't run into this problem, but even in the case where a model checkpoint from a different training session is being finetuned in a new training session with a different machine, do we envision that world size would be different (for our use cases at least we have config files that keep constant variables like Trainer(gpus) for the same model training)?

Comment (justusschock) [reply pytorch_lightning/plugins/training_type/fully_sharded.py:191]: I think that is definitely something we need to support. For example I usually use the output from `torch.cuda.device_count()` as number of gpus.

File: pytorch_lightning/plugins/training_type/fully_sharded.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/plugins/training_type/fully_sharded.py]: In this case, I don't see a problem in it. The only thing I am afraid of, is that we may have a precedence case here and that might lead to this pattern where we don't want it.

What do you think @SeanNaren @awaelchli @ananthsub @carmocca ?

File: pytorch_lightning/plugins/training_type/fully_sharded.py [hunk 9]
```diff
@@ -178,6 +182,25 @@ def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
         # state dict.
         return super().lightning_module_state_dict()
 
+    def serialized_restore_model_state(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
+        checkpoint = {}
+        rank_zero_info(
+            f"FullyShardedDataParallel has {self.num_processes} processes. Serializing model
+            state restore to avoid CPU OOMs."
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/fully_sharded.py:189]:
```suggestion
            state restoration to avoid CPU OOMs."
```

File: pytorch_lightning/plugins/training_type/fully_sharded.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/plugins/training_type/fully_sharded.py:191]: @tchaton this kind of training "metadata" should get saved with the checkpoint. For example, we will also want to know this for fault-tolerance to fail if the trainer configuration has changed between runs and the user is trying to restore mid-batch.

Comment (mleshen) [reply pytorch_lightning/plugins/training_type/fully_sharded.py:191]: FWIW i ran a test taking a checkpoint from a previous training session and starting a fresh one: setting trainer.resume_from_checkpoint to the old checkpoint and trainer.gpus=4 instead of 8 (which was num gpus on the original training session) and loading the checkpoint on 4 GPUs didn't break. 

i can add this feature in another pull request! would a good place to add the metadata be `on_save_checkpoint` in `model_checkpoint.py`? eg here: https://github.com/PyTorchLightning/pytorch-lightning/blob/c7f8c8c3c82b4f249125885490b2392bf9d3d08b/pytorch_lightning/callbacks/model_checkpoint.py#L341

**Comment by tchaton** [comment]
Hey @mleshen, any updates on this PR ? Do you need some assistance ?

**Comment by maximsch2** [comment]
@mleshen , and others - this has been an issue for us a few times in the past and seems to be a particularly tricky things to test for. Do we have tests for checkpoint loading that we can amend to include memory tracking to make sure memory used doesn't scale with number of workers?
